### DISCUSSION TITLE
Loading fields from unloaded classes?

### DISCUSSION THREAD

**Discussion by snightshade** [classgraph/classgraph#790] [Q&A]
I'd like to know the best way to load fields from classes that are not loaded at scan-time. I'm running into the problem of needing to scan the entire classpath, and would like to discover annotated fields in my own classes. I'm running inside `fabric-loader` (the Minecraft modding platform).

Here's the code that does the scanning:
```kotlin
    fun scanTypes() {
        Log.info(LogCategory.LOG, "Scanning types...")
        val launcher = FabricLauncherBase.getLauncher()
        val classLoader = launcher.targetClassLoader
        val classPath = launcher.classPath

        val scanResult = ClassGraph()
                .overrideClassLoaders(classLoader)
                .overrideClasspath(classPath)
                .enableAllInfo()
                .ignoreFieldVisibility()
                .acceptPackages("network.nightshade.myPackage")
                .scan()

        val standards = scanResult.allStandardClasses

        Log.info(LogCategory.LOG, "Standards size = ${standards.size}")

        standards.forEach {
            it.declaredFieldInfo.forEach { that ->
                val info = that.getAnnotationInfo(ConfigKey::class.java)
                if (info != null) {
                    keys[info.name] = that.loadClassAndGetField()
                }
            }
        }

        Log.info(LogCategory.LOG, "Done! Found ${keys.size} types.")
    }
```

My problem is that `it.fieldInfo` (and hence `declaredFieldInfo`) is `null` in the IntelliJ IDEA debugger, causing it to be an empty array, so something is clearly not loading.

The field
```kt
@ConfigKey("some.key")
val field: SerializableBoolean = SerializableBoolean()
```
should end up within `keys` as `{ some.key: <theField> }`.

**Comment by lukehutch** [comment]
ClassGraph never loads classes, ever, unless that is requested (the `.loadClass...` methods, as you are using).

So if `it.fieldInfo` is `null`, that doesn't indicate a class is not being loaded when it is supposed to be loaded (classes aren't loaded by ClassGraph by default). What it means is that when the classfile representing the `ClassInfo` called `it` is scanned, no fields were found.

I don't know the reason for this, but you can call `.verbose()` before `.scan()` to try to figure out why.
